File: internal/runtime/trust/trust.go
// Package trust enforces explicit user approval of distro boot scripts.
//
// Boot scripts are arbitrary Python executed unconditionally on kernel start.
// A trust DB at $TABULA_HOME/state/trust.json records the SHA256 of every
// distro source tree the user has approved. When the kernel boots, it
// recomputes the hash and compares; if the entry is missing or stale, it
// refuses to start and tells the operator how to re-trust.
//
// SHA scope: every *.py file plus distro.toml, recursively, deterministic
// order. Whitespace-only edits to any tracked file invalidate the trust.
// Files outside that filter (markdown docs, templates, prompts) are excluded
// because they are inert from a code-execution standpoint.
package trust

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// Record is one entry in the trust DB.
type Record struct {
	BootSHA256 string `json:"boot_sha256"`
	TrustedAt  string `json:"trusted_at"`
	TrustedBy  string `json:"trusted_by"`
}

// DB is the on-disk shape of $TABULA_HOME/state/trust.json.
type DB struct {
	Distros map[string]Record `json:"distros"`
}

// Meta is the on-disk shape of $TABULA_HOME/state/trust.meta.json. It is
// only used to record that the migration shim auto-trusted an existing
// installation, so the shim can be removed safely in a future release.
type Meta struct {
	MigratedAt string `json:"migrated_at,omitempty"`
}

// Load reads ``path`` and returns the parsed DB. A missing file yields an
// empty DB without error — callers want the same shape regardless.
func Load(path string) (DB, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return DB{Distros: map[string]Record{}}, nil
		}
		return DB{}, fmt.Errorf("read trust db %s: %w", path, err)
	}
	var db DB
	if len(data) == 0 {
		return DB{Distros: map[string]Record{}}, nil
	}
	if err := json.Unmarshal(data, &db); err != nil {
		return DB{}, fmt.Errorf("parse trust db %s: %w", path, err)
	}
	if db.Distros == nil {
		db.Distros = map[string]Record{}
	}
	return db, nil
}

// Save writes ``db`` to ``path`` atomically (temp + rename) so a crash
// cannot leave a half-written file.
func Save(path string, db DB) error {
	if db.Distros == nil {
		db.Distros = map[string]Record{}
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return fmt.Errorf("trust db parent dir: %w", err)
	}
	payload, err := json.MarshalIndent(db, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal trust db: %w", err)
	}
	payload = append(payload, '\n')
	tmp, err := os.CreateTemp(filepath.Dir(path), filepath.Base(path)+".*.tmp")
	if err != nil {
		return fmt.Errorf("trust db tempfile: %w", err)
	}
	tmpName := tmp.Name()
	defer func() { _ = os.Remove(tmpName) }()
	if _, err := tmp.Write(payload); err != nil {
		_ = tmp.Close()
		return fmt.Errorf("write trust db: %w", err)
	}
	if err := tmp.Sync(); err != nil {
		_ = tmp.Close()
		return fmt.Errorf("sync trust db: %w", err)
	}
	if err := tmp.Close(); err != nil {
		return fmt.Errorf("close trust db: %w", err)
	}
	if err := os.Rename(tmpName, path); err != nil {
		return fmt.Errorf("rename trust db: %w", err)
	}
	return nil
}

// HashDir computes a deterministic SHA256 over every ``*.py`` file and the
// top-level ``distro.toml`` inside ``dir``. Paths are recorded relative to
// ``dir`` so renames of the distro home directory do not invalidate trust.
//
// Layout of the hashed stream, for each file in sorted-relative-path order:
//
//	"<rel-path>\x00<len>\x00<bytes>\n"
//
// The ``len`` prefix prevents a payload from spilling into the next entry
// if someone manages to slip a NUL into a filename. Symlinks are skipped.
// Files under ``__pycache__``, ``.git``, and any dot-prefixed directory
// are skipped because they are derived or unrelated artifacts.
func HashDir(dir string) (string, error) {
	info, err := os.Stat(dir)
	if err != nil {
		return "", fmt.Errorf("trust hash: stat %s: %w", dir, err)
	}
	if !info.IsDir() {
		return "", fmt.Errorf("trust hash: %s is not a directory", dir)
	}
	var entries []string
	err = filepath.WalkDir(dir, func(path string, d os.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if d.IsDir() {
			name := d.Name()
			if path != dir && (name == "__pycache__" || strings.HasPrefix(name, ".")) {
				return filepath.SkipDir
			}
			return nil
		}
		if !d.Type().IsRegular() {
			return nil
		}
		name := d.Name()
		rel, err := filepath.Rel(dir, path)
		if err != nil {
			return err
		}
		// distro.toml only matters at the top of the tree. Anywhere else
		// (a fixture or test asset) we leave it alone.
		if name == "distro.toml" && filepath.Dir(rel) == "." {
			entries = append(entries, rel)
			return nil
		}
		if strings.HasSuffix(name, ".py") {
			entries = append(entries, rel)
		}
		return nil
	})
	if err != nil {
		return "", fmt.Errorf("trust hash: walk %s: %w", dir, err)
	}
	sort.Strings(entries)

	h := sha256.New()
	for _, rel := range entries {
		full := filepath.Join(dir, rel)
		data, err := os.ReadFile(full)
		if err != nil {
			return "", fmt.Errorf("trust hash: read %s: %w", full, err)
		}
		// Normalise rel to forward slashes so a Windows install computes
		// the same hash as a Unix one. We do not currently ship Windows
		// but the cost is one strings.ReplaceAll.
		relSlash := filepath.ToSlash(rel)
		if _, err := io.WriteString(h, relSlash); err != nil {
			return "", err
		}
		if _, err := io.WriteString(h, "\x00"); err != nil {
			return "", err
		}
		if _, err := fmt.Fprintf(h, "%d", len(data)); err != nil {
			return "", err
		}
		if _, err := io.WriteString(h, "\x00"); err != nil {
			return "", err
		}
		if _, err := h.Write(data); err != nil {
			return "", err
		}
		if _, err := io.WriteString(h, "\n"); err != nil {
			return "", err
		}
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

// Check returns nil when ``dir`` matches the trust DB entry for
// ``distroID``. The returned error is always wrapped with one of the
// sentinel values below so callers can branch on intent.
//
// Error semantics:
//
//   - ErrNotTrusted   — no entry exists; user must run `tabula distro trust`.
//   - ErrSHAMismatch  — entry exists but the distro changed.
//   - other errors    — I/O or hash failures, surface unchanged.
func Check(dbPath, distroID, dir string) error {
	if strings.TrimSpace(distroID) == "" {
		return fmt.Errorf("trust check: empty distro id")
	}
	if strings.TrimSpace(dir) == "" {
		return fmt.Errorf("trust check: empty distro dir")
	}
	db, err := Load(dbPath)
	if err != nil {
		return err
	}
	record, ok := db.Distros[distroID]
	if !ok {
		return &CheckError{Kind: ErrNotTrusted, DistroID: distroID, Dir: dir}
	}
	actual, err := HashDir(dir)
	if err != nil {
		return err
	}
	if actual != record.BootSHA256 {
		return &CheckError{
			Kind:     ErrSHAMismatch,
			DistroID: distroID,
			Dir:      dir,
			Expected: record.BootSHA256,
			Actual:   actual,
		}
	}
	return nil
}

// Approve writes (or refreshes) the trust record for ``distroID`` after
// computing the current hash of ``dir``. ``trustedBy`` is a free-form
// label, typically "user" or "installer-cold-start".
func Approve(dbPath, distroID, dir, trustedBy string) (Record, error) {
	digest, err := HashDir(dir)
	if err != nil {
		return Record{}, err
	}
	db, err := Load(dbPath)
	if err != nil {
		return Record{}, err
	}
	if db.Distros == nil {
		db.Distros = map[string]Record{}
	}
	record := Record{
		BootSHA256: digest,
		TrustedAt:  time.Now().UTC().Format(time.RFC3339),
		TrustedBy:  trustedBy,
	}
	db.Distros[distroID] = record
	if err := Save(dbPath, db); err != nil {
		return Record{}, err
	}
	return record, nil
}

// Revoke deletes the trust record for ``distroID``. Missing entries are a
// no-op.
func Revoke(dbPath, distroID string) error {
	db, err := Load(dbPath)
	if err != nil {
		return err
	}
	if _, ok := db.Distros[distroID]; !ok {
		return nil
	}
	delete(db.Distros, distroID)
	return Save(dbPath, db)
}

// ErrKind enumerates trust-check failure categories. Callers use it via
// errors.As(err, &*CheckError{}) and switch on Kind.
type ErrKind int

const (
	// ErrNotTrusted means the trust DB has no record for the distro.
	ErrNotTrusted ErrKind = iota + 1
	// ErrSHAMismatch means the recorded SHA does not match the on-disk one.
	ErrSHAMismatch
)

// CheckError is the structured error returned by Check.
type CheckError struct {
	Kind     ErrKind
	DistroID string
	Dir      string
	Expected string
	Actual   string
}

func (e *CheckError) Error() string {
	switch e.Kind {
	case ErrNotTrusted:
		return fmt.Sprintf(
			"distro %q has not been trusted; run `tabula distro trust %s` after reviewing %s",
			e.DistroID, e.DistroID, e.Dir,
		)
	case ErrSHAMismatch:
		return fmt.Sprintf(
			"distro %q boot tree changed since last trust\n  expected SHA: %s\n  actual SHA:   %s\n  run `tabula distro trust %s` to approve",
			e.DistroID, e.Expected, e.Actual, e.DistroID,
		)
	default:
		return fmt.Sprintf("trust check failed for %q", e.DistroID)
	}
}

File: tools/tabula-distro/src/tabula_distro/trust.py
"""Trust DB helper for the installer side.

The kernel reads ``$TABULA_HOME/state/trust.json`` to gate boot execution
(see ``tabula/internal/runtime/trust``). The installer needs to write the
same file: explicit ``--trust`` after a fresh install, and the cold-start
migration shim that auto-trusts an existing installation upgraded from
before issue 007.

Hash algorithm matches the Go side byte for byte:

    sorted-relative-path-newline-separated stream of
    "<rel>\0<len>\0<bytes>\n" entries, SHA256.

Files included: ``*.py`` recursively, plus top-level ``distro.toml``.
Excluded: hidden directories (``.git``, ``.venv``, …), ``__pycache__``,
non-regular files (symlinks/sockets/devices), and ``distro.toml`` outside
the top of the tree.
"""

from __future__ import annotations

import hashlib
import json
import os
import tempfile
from datetime import datetime, timezone
from pathlib import Path

from tabula_plugin_sdk import paths as sdk_paths


class TrustError(RuntimeError):
    """Raised on trust DB read/write failures."""


def trust_db_path(home: Path) -> Path:
    return home / "state" / "trust.json"


def trust_meta_path(home: Path) -> Path:
    return home / "state" / "trust.meta.json"


def hash_dir(distro_dir: Path) -> str:
    """Return SHA256 over the distro tree.

    Mirrors ``internal/runtime/trust.HashDir`` in Go. Any drift in this
    function silently invalidates every existing trust record, so changes
    here must come paired with a kernel-side update and a regen of every
    user's trust DB.
    """
    distro_dir = distro_dir.expanduser()
    if not distro_dir.is_dir():
        raise TrustError(f"distro dir does not exist: {distro_dir}")

    entries: list[str] = []
    for root, dirs, files in os.walk(distro_dir):
        # Prune walks before recursing so we never read excluded files.
        dirs[:] = [
            d for d in dirs
            if d != "__pycache__" and not d.startswith(".")
        ]
        for name in files:
            full = Path(root) / name
            if not full.is_file() or full.is_symlink():
                continue
            rel = full.relative_to(distro_dir)
            rel_posix = rel.as_posix()
            if name == "distro.toml":
                # Only the top-level distro.toml is part of the contract.
                if rel.parent == Path("."):
                    entries.append(rel_posix)
                continue
            if name.endswith(".py"):
                entries.append(rel_posix)
    entries.sort()

    h = hashlib.sha256()
    for rel in entries:
        data = (distro_dir / rel).read_bytes()
        h.update(rel.encode("utf-8"))
        h.update(b"\x00")
        h.update(str(len(data)).encode("ascii"))
        h.update(b"\x00")
        h.update(data)
        h.update(b"\n")
    return h.hexdigest()


def load(home: Path) -> dict:
    path = trust_db_path(home)
    if not path.is_file():
        return {"distros": {}}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise TrustError(f"parse trust db {path}: {exc}") from exc
    if not isinstance(data, dict):
        raise TrustError(f"trust db {path} is not a JSON object")
    if not isinstance(data.get("distros"), dict):
        data["distros"] = {}
    return data


def save(home: Path, db: dict) -> Path:
    path = trust_db_path(home)
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = json.dumps(db, indent=2, sort_keys=True) + "\n"
    fd, tmp_name = tempfile.mkstemp(
        prefix=path.name + ".",
        suffix=".tmp",
        dir=str(path.parent),
    )
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            fh.write(payload)
            fh.flush()
            os.fsync(fh.fileno())
        os.replace(tmp_name, path)
    except BaseException:
        try:
            os.unlink(tmp_name)
        except OSError:
            pass
        raise
    return path


def approve(home: Path, distro_id: str, distro_dir: Path, *, trusted_by: str) -> dict:
    """Compute the SHA and write a trust record. Returns the record."""
    digest = hash_dir(distro_dir)
    db = load(home)
    record = {
        "boot_sha256": digest,
        "trusted_at": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "trusted_by": trusted_by,
    }
    db["distros"][distro_id] = record
    save(home, db)
    return record


def auto_trust_cold_start(home: Path, distro_id: str, distro_dir: Path) -> dict | None:
    """Migration shim: trust an existing install if no trust DB exists yet.

    Returns the recorded record on first run, ``None`` on subsequent runs.

    The shim is intentionally one-shot: it only fires when
    ``state/trust.json`` is absent, so a user who deliberately revoked
    trust does not get silently re-approved. After it fires we drop a
    ``state/trust.meta.json`` marker with the migration timestamp so the
    shim can be removed cleanly in a future release.

    See ``docs/issues/refactoring/007-distro-boot-trust-db.md`` § Risk and
    Migration.
    """
    if trust_db_path(home).exists():
        return None
    record = approve(home, distro_id, distro_dir, trusted_by="installer-cold-start")
    meta_path = trust_meta_path(home)
    meta_path.parent.mkdir(parents=True, exist_ok=True)
    meta_path.write_text(
        json.dumps(
            {"migrated_at": record["trusted_at"]},
            indent=2,
            sort_keys=True,
        )
        + "\n",
        encoding="utf-8",
    )
    return record


# Re-export the trust file path through the shared paths helper so distro
# tooling can reach for it the same way the kernel does. The actual file
# definition lives here because the installer needs writer semantics that
# the read-only kernel does not.
__all__ = [
    "TrustError",
    "trust_db_path",
    "trust_meta_path",
    "hash_dir",
    "load",
    "save",
    "approve",
    "auto_trust_cold_start",
]


def _consistency_check_paths_module() -> None:
    """Belt-and-braces: assert the SDK paths module agrees on the file name.

    Only invoked from tests via ``__main__`` so production startup is not
    slowed by it.
    """
    home = sdk_paths.tabula_home()
    if trust_db_path(home) != sdk_paths.trust_file():
        raise TrustError(
            f"trust_db_path mismatch: {trust_db_path(home)} vs {sdk_paths.trust_file()}",
        )
